Add git function tutorial (#9379)
kyma-project/kyma: docs/serverless/08-02-create-git-function.md

---

> title: Create a Function from Git Repository

Reviewer: ```suggestion
title: Create a Function from Git repository sources
```

Author: what about:
```suggestion
title: Create a Function from remote Git repository
```?

type: Tutorials
---

This tutorial shows how you can create a sample Function from Git repository based on [orders service example](https://github.com/kyma-project/examples/tree/master/orders-service).

## Steps

Follows these steps:

<div tabs name="steps" group="create-function">
  <details>
  <summary label="cli">
  CLI
  </summary>

1. Export these variables:

    ```bash
    export NAME={FUNCTION_NAME}
    export NAMESPACE={FUNCTION_NAMESPACE}
    ```

    (Optional) If you are willing to use a private repository, create a secret with credentials to the repository:

    ```yaml
    cat <<EOF | kubectl apply -f -
    apiVersion: v1
    kind: Secret
    metadata:
      name: git-creds-basic
      namespace: $NAMESPACE
    type: Opaque
    data:
      username: <USERNAME>
      password: <PASSWORD>
    EOF
    ```

>     >**NOTE** To see other authorization methods, go to the [documentation]().

Author: Ok, but still I would like to mention right below that example there are other authentication methods that we support. Maybe something like 
```suggestion
    >**NOTE** Check the documentation to see other supported authentication methods.
```

?

2. Create a GitRepository CR that specifies the Git repository metadata:

    ```yaml
    cat <<EOF | kubectl apply -f -
    apiVersion: serverless.kyma-project.io/v1alpha1
    kind: GitRepository
    metadata:
      name: $NAME
      namespace: $NAMESPACE
    spec:
      url: "https://github.com/kyma-project/examples.git"
    EOF
    ```
   
    >**NOTE** If you are using a private repository, add `auth` object with `type` and `secretName` fields to the spec. For details, see [documentation]().

> 3. Create a Function CR that specifies the Function's logic:

Author: what about:
```suggestion
3. Create a Function CR that specifies the Function's logic and points to the directory with code and dependencies inside the given repository
```

?

    ```yaml
    cat <<EOF | kubectl apply -f -
    apiVersion: serverless.kyma-project.io/v1alpha1
    kind: Function
    metadata:
      name: $NAME
      namespace: $NAMESPACE
    spec:
      type: git
      runtime: nodejs12
      source: $NAME
      reference: master
      baseDir: orders-service/function
    EOF
    ```

    >**NOTE** To see the function files, go to [this](https://github.com/kyma-project/examples/tree/master/orders-service/function) page. 

4. Check if your Function was created successfully and all conditions are set to `True`:

    ```bash
    kubectl get functions $NAME -n $NAMESPACE
    ```

    You should get a result similar to the following example:

    ```bash
    NAME                        CONFIGURED   BUILT   RUNNING   VERSION   AGE
    test-function               True         True    True      1         18m
    ```

    </details>
    <details>
    <summary label="console-ui">
    Console UI
    </summary>

> 1. Create a Namespace or select one from the drop-down list in the top navigation panel.

Reviewer: A bit better, but it still needs updating:
![image](https://user-images.githubusercontent.com/40655785/92724755-69c83480-f36b-11ea-93c4-c86d827394c2.png)

    (Optional) If you are willing to use a private repository, to authorize with basic authentication create a secret with credentials to the repository:

    1. Save yaml file with secret on your hard drive:
    ```yaml
    apiVersion: v1
    kind: Secret
    metadata:
      name: git-creds-basic
    type: Opaque
    data:
      username: <USERNAME>
      password: <PASSWORD>
    ```

>    >**NOTE** To see other possible authorization methods, go to the [documentation]().

Author: Again, as with the CLI example, I'd like to mention that we support more than one (basic in that specific case) authorization method

   
   2. Go to the namespace view and click **Deploy new resource**.
   3. Select secret file and click **Deploy**.

2. Go to the **Functions** view in the left navigation panel and select **Repositories** tab.

3. Click **Connect Repository**, fill the `Url` field with `https://github.com/kyma-project/examples.git` value and click **Connect**.

    >**NOTE** If you want connect the private repository, change the `Authorization` field from `Public` to `Key` or `Basic` and fill the required fields. For details, see [documentation]().

4. Go to the **Functions** tab and click **Create Function**.

> 5. In the pop-up box, change `Source type` to `From Repository`, select created Repository's name, fill the `Reference` field with `master` and `Base directory` field with `orders-service/function` values and select **Create** to confirm changes.

Reviewer: (I asked Maciek and it will be capitalized - as in other views)

    The pop-up box closes and the message appears on the screen after a while, confirming that the Function was created successfully.

>     The new Function should have the `RUNNING` status in the list of all Functions under the **Functions** view.

Reviewer: ```suggestion
    Make sure that new Function has the `RUNNING` status in the list of all Functions under the **Functions** view.
```

---
title: Create a Function from Git repository sources
type: Tutorials
---

Create a sample Function from code and dependencies stored in a Git repository. The tutorial is based on the Function from the [`orders service` example](https://github.com/kyma-project/examples/tree/master/orders-service). It shows how you can fetch Function's source code and dependencies from a public Git repository that does not require any authentication method.

> **NOTE:** Read more about [Git source type](#details-git-source-type) and different ways of securing your repository.

## Steps

Follows these steps:

<div tabs name="steps" group="create-function">
  <details>
  <summary label="kubectl">
  kubectl
  </summary>

> 1. Export these variables:

Reviewer: Sth is wrong with formatting: 
![image](https://user-images.githubusercontent.com/40655785/92724631-41403a80-f36b-11ea-878e-91bdef67fe2e.png)

    
    1.1. On your machine, create this YAML file with the Secret definition:
    
    ```yaml
    apiVersion: v1
    kind: Secret
    metadata:
      name: git-creds-basic
    type: Opaque
    data:
      username: <USERNAME>
      password: <PASSWORD>
    ```
    1.2. Go to your Namespace view and select **Deploy new resource**. 
        
    1.3. Locate the YAML file with the Secret and select **Deploy**.
   
    >**NOTE:** Read also about other [supported authentication methods](#details-git-source-type).
   
2. Go to the **Functions** view in the left navigation panel and select the **Repositories** tab.

3. Select **Connect Repository**, fill in the **Url** field with `https://github.com/kyma-project/examples.git`, and confirm by selecting **Connect**.

    >**NOTE:** If you want to connect the secured repository with basic authentication, change the **Authorization** field from `Public` to `Basic` and fill in the required fields.

4. Go to the **Functions** tab and select **Create Function**.

> 5. In the pop-up box, change `Source Type` to `From Repository`, select created Repository's name, fill the `Reference` field with `master` and `Base Directory` field with `orders-service/function` values and select **Create** to confirm changes.

Reviewer: ```suggestion
5. In the pop-up box, change `Source Type` to `From Repository`, select created Repository's name, fill in the `Reference` field with `master` and `Base Directory` field with `orders-service/function` values, and select **Create** to confirm changes.
```

---
title: Create a Function from Git Repository
type: Tutorials
---

> This tutorial shows how you can create a simple "Hello World!" Function from Git repository.

Reviewer 2: `This tutorial shows how you can create a Function from Git repository.` The function from `examples` repo is a more complex than `Hello World!` :)

## Steps

Follows these steps:

<div tabs name="steps" group="create-function">
  <details>
  <summary label="cli">
  CLI
  </summary>

1. Export these variables:

    ```bash
    export NAME={FUNCTION_NAME}
    export NAMESPACE={FUNCTION_NAMESPACE}
    ```

2. Create a GitRepository CR that specifies the Git repository metadata:

    ```yaml
    cat <<EOF | kubectl apply -f -
    apiVersion: serverless.kyma-project.io/v1alpha1
    kind: GitRepository
    metadata:

>       name: $NAME

Reviewer 2: Add also namespace `namespace: $NAMESPACE`

    spec:

>       url: "https://github.com/pPrecel/public-gitops"

Reviewer 2: Please use the https://github.com/kyma-project/examples.git repository. It's better to ref to repo in organisation, not to repo in account one of our contributors.

    EOF
    ```
   

> >**NOTE** Auth possibility TODO

Reviewer 2: Please add correct indents (3 or 4 spaces)

3. Create a Function CR that specifies the Function's logic:

    ```yaml
    cat <<EOF | kubectl apply -f -
    apiVersion: serverless.kyma-project.io/v1alpha1
    kind: Function
    metadata:

>       name: $NAME

Reviewer 2: Add also namespace `namespace: $NAMESPACE`

    spec:
      type: git
      runtime: nodejs12
      source: $NAME
      reference: master

>       baseDir: js-handler

Reviewer 2: Change to `orders-service/function`

    EOF
    ```

>     >**NOTE** To see full spec, go to the xxx page 

Reviewer 2: Go to https://github.com/kyma-project/examples/tree/master/orders-service/function

3. Check if your Function was created successfully and all conditions are set to `True`:

    ```bash
    kubectl get functions $NAME -n $NAMESPACE
    ```

    You should get a result similar to the following example:

    ```bash
    NAME                        CONFIGURED   BUILT   RUNNING   VERSION   AGE
    test-function               True         True    True      1         18m
    ```

    </details>
    <details>
    <summary label="console-ui">
    Console UI
    </summary>

1. Create a Namespace or select one from the drop-down list in the top navigation panel.

2. Go to the **Functions** view in the left navigation panel and select **Repositories** tab.

> 3. Click **Connect Repository**,  fill in required fields and click **Connect**.

Reviewer 2: Click **Connect Repository**, fill the `Url` field with `https://github.com/kyma-project/examples.git` value and click **Connect**.

>     >**NOTE** Auth possibility TODO

Reviewer 2: > **NOTE** If you want connect the private repository, change the `Authorization` field from `Public` to `Key` or `Basic` and fill the required fields -> Auth possibility link TODO

4. Go to the **Functions** tab and click **Create Function**.

> 3. In the pop-up box, change `Source type` to `From Repository`, fill in required fields and select **Create** to confirm changes.

Reviewer 2: 5. In the pop-up box, change `Source type` to `From Repository`, select created Repository's name, fill the `Reference` field with `master` and `Base directory` field with `orders-service/function` values and select **Create** to confirm changes.
